Migrate forking to a pooled worker design #1428

ijjk opened the pull request:
Here's the start of my implementation for migrating forking to a pooled worker design. My implementation has a `SingleProcessTestPool`, `SharedForkTestPool`, and `ForkTestPool` which share a common interface to try to help abstract away the method of executing. 

~~Currently, it picks a `TestPool`  to use based on the factors set out on the issue~~:

- ~~For <N tests (where N is the number of processors), use SingleProcessTestPool; otherwise, use ForkTestPool~~
- ~~When --no-fork or equivalent flag is provided, force SingleProcessTestPool~~
- ~~When --inspect or --inspect-brk is specified, force SingleProcessTestPool~~

~~In the thread, the `ForkTestPool` was described as distributing the tests to the different forks as they become idle, and I thought it might simplify the execution to divide the tests beforehand and then create the forks to run them. Doing it this way allowed me to re-use the `SingleProcessTestPool` implementation inside of the forked process~~.

~~To achieve the testing inside the single process I updated the import of the `runner` to use a map with the test file as the key and the runner as the value. I also, utilized `vm` to provide isolation between running the tests in the same thread~~. Fixes: #1428 Fixes: #1332 

After review by @novemberborn, there are now no breaking changes and use of the new test pools are opt-in. Also, the `SingleProcessTestPool` does not run inside of a sandbox e.g. `vm` as to avoid having to maintain a sandboxing library since `vm2` does not work for this context. You can opt-in to a test pool with below configs/flags

- `SingleProcessTestPool` using the `--single-process` flag or `singleProcess` config
- `SharedForkTestPool` using the `--share-forks` flag or `shareForks` config

You can also make use of `worker_threads` now on supported node versions (>= 11.8 or > 10.x with `--experimental-worker` node flag) using the `--worker-threads` flag or `workerThreads` config

TODO:
- [x] Add note for `--no-fork` flag and `fork` config in docs
- [x] Clean up timeouts
- [x] Fix failing test cases

novemberborn commented:
Hey @ijjk thanks for picking this up! There's a lot of changes here so I may not be able to have a proper look until next weekend. Going by the commit messages it looks like you're really far along already!

ijjk commented:
@novemberborn okay thanks for the reply! Yeah, the changes started to add up. I'm still sorting out some test cases, and then hopefully it will be good to go.

lo1tuma commented:
I’ve tried to run this against the test-suite of one of our projects with 1608 tests in total. While ava v1 takes round about 37 seconds, with this changes it seems to be extremely fast (my gut feeling < 5 seconds with the ForkTestPool). The problem is that after `1222` tests have been completed the runner completely hangs and I have to manually abort to process. I see some warnings on the console:

```
 MaxListenersExceededWarning: Possible EventEmitter memory leak detected. 25 unhandledRejection listeners added. Use emitter.setMaxListeners() to increase limit
```

The same happens when I use the `--no-fork` option.

Apart from that the reporting seems to behave strange, does it show a spinner for every process? Here is  a screenshot:
<img width="825" alt="screen shot 2019-01-23 at 16 53 18" src="https://user-images.githubusercontent.com/169170/51619400-89111480-1f30-11e9-8f57-9ac9282a738b.png">

As you can see in the screenshot there is also a warning from react which complains about missing `requestAnimationFrame`. I don’t get this warning with ava v1. I would guess this is related to sandboxing. In some test files I set `global.window` to an instance of `jsdom`.

As a follow up feature it would be also nice to provide an option which disables sandboxing in favor of performance.

novemberborn commented on index.js:
```diff
@@ -1,8 +1,21 @@
+const {filename} = module.parent;
+
+if (filename.indexOf('node_modules/ava') > -1) {
+	// If request was routed by a different ava install to
```
`AVA` please 😄

novemberborn commented on index.js:
```diff
@@ -1,8 +1,21 @@
 	module.exports = require('./lib/worker/main');
+} else if (process.env.AVA_PATH && process.env.AVA_PATH !== __dirname) {
+	// Ensure the same AVA install is loaded by the test file as by the test worker
+	module.exports = require(process.env.AVA_PATH)();
```
In the original code, the required module is not invoked. Why is it invoked here?

novemberborn commented on index.js:
```diff
@@ -1,8 +1,21 @@
+	// Ensure the same AVA install is loaded by the test file as by the test worker
+	module.exports = require(process.env.AVA_PATH)();
+} else {
+	module.exports = require('./lib/worker/main')();
```
Should the `main` module be invoked?

novemberborn commented on lib/babel-pipeline.js:
```diff
@@ -44,6 +43,8 @@ function validate(conf) {
 		return {testOptions: defaultOptions};
 	}
 
+	const chalk = require('./chalk').get();
```
Why did this need moving?

ijjk replied:
This was left over from some previous debugging sorry

novemberborn commented on profile.js:
```diff
@@ -69,97 +36,9 @@ if (cli.input.length === 0) {
+if (importLocal(__filename)) {
+	debug('Using local install of AVA');
+} else {
+	require('./lib/cli').run({fork: false});
```
I think we can remove `profile.js` with this PR. It's pretty awkward and you could now run `npx ava --no-fork --inspect my-test.js` instead.

Then we can also remove the `overrideConfig` from the `lib/cli`.

ijjk replied:
I was thinking it could be removed also. I removed it and updated the docs with the new command.

novemberborn commented on lib/fork.js:
```diff
@@ -1,132 +1,107 @@
-			emitStateChange({type: 'worker-failed', err});
-			finish();
-		});
+const failTypes = {
```
Use a `Set` instead.

novemberborn commented on lib/fork.js:
```diff
@@ -1,132 +1,107 @@
+			// Listen for messages
+			this.subprocess.on('message', msg => {
+				if (msg.forkMsg === 'new-file') {
+					runStatus.observeWorker({onStateChange: () => {}}, msg.file);
```
Looks like `run-status` needs updating?

ijjk replied:
After re-working the branch, I don't think `run-status` need updating anymore. Do you think it still does?

novemberborn commented on lib/fork.js:
```diff
@@ -1,132 +1,107 @@
+					runStatus.emitStateChange(msg);
+					// Check if we need to trigger exits
+					if (failFast && failTypes[msg.type]) {
+						reject(new Error('notify-fail'));
```
This should be decided higher-up I think.

novemberborn commented on lib/testpool/fork-test-pool.js:
```diff
@@ -0,0 +1,118 @@
+class ForkTestPool extends TestPool {
+	bail() {
+		debug('bailing forks');
+		this.forks.map(fork => fork.exit());
```
Use a `for`/`of` loop.

novemberborn commented on lib/testpool/fork-test-pool.js:
```diff
@@ -0,0 +1,118 @@
+		return Bluebird.map(fileChunks, curFiles => {
+			return this._computeForkExecArgv().then(execArgv => {
+				this.forks.push(
+					new Fork(curFiles, runStatus, this.env, this.initOptions, execArgv)
```
This… gives a single `Fork` the responsibility to run multiple test files? I can't quite tell how `Fork` implements this but it also seems a little odd. You could end up with a slow fork that still has two pending files while other slots are idle.

novemberborn commented on lib/testpool/fork-test-pool.js:
```diff
@@ -0,0 +1,118 @@
+			return Bluebird.map(this.forks, fork => {
+				return fork.run().catch(() => {
+					if (failFast) {
+						this.forks.map(fork => fork.exit());
```
Use `for`/`of`.

novemberborn commented on lib/testpool/fork-test-pool.js:
```diff
@@ -0,0 +1,118 @@
+					if (failFast) {
+						this.forks.map(fork => fork.exit());
+						return Promise.reject(new Error('clean-exit'));
+					}
```
Need to rethrow other errors.

novemberborn commented on lib/testpool/fork-test-pool.js:
```diff
@@ -0,0 +1,118 @@
+				return fork.run().catch(() => {
+					if (failFast) {
+						this.forks.map(fork => fork.exit());
+						return Promise.reject(new Error('clean-exit'));
```
You can throw, no need to return a rejected promise.

novemberborn commented on lib/testpool/fork-test-pool.js:
```diff
@@ -0,0 +1,118 @@
+		})
+			.catch(err => {
+				// Exit cleanly if it's handled, if not re-throw
+				if (err.message === 'clean-exit') {
```
I'm not a fan of this approach. I'd rather build on the state change notification mechanism to then interact with specific forks or the pool.

novemberborn commented on lib/testpool/fork-test-pool.js:
```diff
@@ -0,0 +1,118 @@
+					return;
+				}
+
+				throw new Error(err);
```
`throw err`

novemberborn commented on lib/testpool/single-process-test-pool.js:
```diff
@@ -0,0 +1,135 @@
+const postTestPath = require.resolve('./test-post-require.js');
+
+const preTestContent = fs.readFileSync(preTestPath, 'utf8').toString();
+const postTestContent = fs.readFileSync(postTestPath, 'utf8').toString();
```
Since you provided the `'utf8'` encoding `readFileSync()` already returns strings.

ijjk replied:
This ended up getting removed, but good to know.

novemberborn commented on lib/testpool/single-process-test-pool.js:
```diff
@@ -0,0 +1,135 @@
+
+	run() {
+		const {options, runOnlyExclusive} = this;
+		const concurrency = options.serial ? 1 : 2;
```
I'm not sure we need to have any test file concurrency within the process. That would also simplify the implementation.

novemberborn commented on lib/testpool/single-process-test-pool.js:
```diff
@@ -0,0 +1,135 @@
+
+		return Bluebird.map(this.files, file => {
+			// Resolve realpath in case it's symlink
+			file = fs.realpathSync(file);
```
I don't think we do this any other time? Is it because when you launch a child process that already takes care of resolving the symlink?

novemberborn commented on lib/testpool/single-process-test-pool.js:
```diff
@@ -0,0 +1,135 @@
+
+			// Check if we are bailing
+			if (this.bailed) {
+				return Promise.resolve();
```
You can just return.

novemberborn commented on lib/testpool/single-process-test-pool.js:
```diff
@@ -0,0 +1,135 @@
+
+			// Let fork know to create file in runStatus
+			if (typeof process.send === 'function') {
+				process.send({forkMsg: 'new-file', file});
```
Why communicate through `process`?

novemberborn commented on lib/testpool/single-process-test-pool.js:
```diff
@@ -0,0 +1,135 @@
+				} catch (err) {
+					let error = err;
+					if (error.message === 'unexpected-exit') {
+						error = {type: 'worker-failed'};
```
Errors should be errors. Let's find a different way of communicating these events.

novemberborn commented on lib/testpool/test-pool.js:
```diff
@@ -0,0 +1,88 @@
+const serializeError = require('../serialize-error');
+
+if (!process.env.NODE_ENV) {
+	process.env.NODE_ENV = 'test';
```
This changes the current behavior. If `process.env.NODE_ENV === ''`, that should be preserved.

novemberborn commented on lib/testpool/test-pool.js:
```diff
@@ -0,0 +1,88 @@
+			evt.testFile = file;
+		}
+
+		// Debug('emitting', evt)
```
?

ijjk commented:
@novemberborn 

> What do you mean by having to use "`require" in the host context"?

`vm2`'s `NodeVM` has the option to require in the `sandbox` or in the `host` context. When using the `sandbox` require mode, modules that need to modify the `module.constructor._extensions` like `@babel/register` fail since `vm2` doesn't expose a full `module` instance. 

You can use the `host` require mode and it works but this implementation breaks pretty much any sandboxing since any changes made globally or to the module on require are done in the host and not the sandbox. 

So for example, in `host` require mode, if you require a module let's call it `browser-setup.js` with below contents, the changes will be made in the `host` context and not the `sandbox` pretty much breaking the isolation.

```js
const { JSDOM } = require('jsdom')
const { window, document } = new JSDOM(`<!DOCTYPE html><p>Hello world</p>`)
global.window = window
global.document = document
global.fetch = require('node-fetch')
```

I couldn't find any existing modules that handle this.

- `vm2` has above problem

- `isolated-vm` runs in a fork and according to the [README](https://github.com/laverdet/isolated-vm#requirements) mainly works with node >= 8.11.2

- `napajs` runs on seperate threads and doesn't support all built-in node_modules

This is why I started working out a bare minimum custom implementation of `vm`.

novemberborn commented:
I see.

What do you think of https://nodejs.org/api/worker_threads.html? That should solve the sandboxing and the performance issues. I still see a need for properly forked processes, so the pool switching will be necessary. Perhaps we could support non-sandboxed single process as an opt-in which only works for some use cases.

ijjk commented:
@novemberborn `WorkerThreads` look interesting although still experimental. I re-worked my branch to support a non-isolated single process mode, the current mode of creating a new fork for each test, and a shared fork mode which reuses the forks and distributes tests to them as they become available. 

The single process mode and shared fork modes are opt-in currently. I could try out a `WorkerThread` implementation that's opt-in too since it's still in the experimental stage. Was this re-work what you had in mind with the previous comment?

novemberborn commented:
> I could try out a `WorkerThread` implementation that's opt-in too since it's still in the experimental stage. Was this re-work what you had in mind with the previous comment?

Yea. It no longer requires a flag in the recent Node.js 11 release(s), so it's definitely an option for us. It'll be in Node.js 12 soon enough!

ijjk commented:
@novemberborn I added the option to use Worker Threads instead of forking. The API was pretty similar so I was able to implement it using `./lib/fork`.

It's also set up so that Worker Threads can be shared in the same way that forks are with the `--share-forks` flag since creating new Worker Threads can be expensive on performance similar to creating new forks. 

On one project's test suite that has around 632 tests across 59 files, that is using AVA, [I found on github](https://github.com/freaktechnik/justintv-stream-notifications), the suite had below run times. Note, by no-cache I mean the cache under `node_modules/.cache` has been cleared for the project.

| Mode | Run duration in seconds |
| ---- | ----------------------- |
| `ForkTestPool` no cache | 48.67s |
| `ForkTestPool` with cache | 38.34s |
| `SharedForkTestPool` no cache | 22.09s |
| `SharedForkTestPool` with cache | 14.91s |
| `WorkerThreads` no cache | 48.43s |
| `WorkerThreads` with cache | 35.76s |
| `SharedWorkerThreads` no cache | 21.51s |
| `SharedWorkerThreads` with cache | 15.60s |
| `SingleProcessTestPool` no cache | Suite can't run in single process |
| `SingleProcessTestPool` with cache | ^ |


I also ran this on [styled-jsx's](https://github.com/zeit/styled-jsx) test-suite which is around 174 tests across 16 files and got below run times.

| Mode | Run duration in seconds |
| ---- | ----------------------- |
| `ForkTestPool` no cache | 10.73s |
| `ForkTestPool` with cache | 6.34s |
| `SharedForkTestPool` no cache | 8.80s |
| `SharedForkTestPool` with cache | 6.30s |
| `WorkerThreads` no cache | 10.84s |
| `WorkerThreads` with cache | 6.96s |
| `SharedWorkerThreads` no cache | 9.17s |
| `SharedWorkerThreads` with cache | 6.37s |
| `SingleProcessTestPool` no cache | 8.11s |
| `SingleProcessTestPool` with cache | 4.36s |

I didn't run these under a super controlled environment so they might not be perfect estimates but thought it might be helpful to compare on actual projects using AVA.

iwilson-r7 commented on lib/test-pools/single-process-test-pool.js:
```diff
@@ -0,0 +1,175 @@
+		const testPath = file;
+
+		// Give runStatus a chance to setup listener
+		setTimeout(() => {
```
I believe this should use `nowAndTimers.setTimeout`, otherwise use of `sinon.useFakeTimers` and forgetting to restore the original timer functions in a test can cause the test run to either hang inexplicably or exit without completing (a difficult to debug problem as I discovered!)

ijjk replied:
Good catch, I actually updated it to use `nowAndTimers.setImmediate` since it just needs to run after `runStatus.observeWorker` has been called.

novemberborn commented:
Hey @ijjk sorry for the delay in getting back to you. It's quite a lot to take in! 

I've left some more comments, and I'm adding a checklist below. I think we may be getting to a point where I need to play around with the code and perhaps make some more changes. If you could point out the parts you're not comfortable with yet that'd be a great help.

Since this is such an involved PR it may take a while before I can really sit down for this. I really appreciate the work you're putting into this though!

Checklist:

* [ ] Review debug() output
* [ ] Review documentation (probably needs a dedicated section / recipe)
* [ ] Bikeshed CLI flags
* [ ] Review error message for when worker threads are unavailable
* [ ] "Shared fork" reusing fork leads to complicated code paths, both in the pool code and in the `Fork` class, the latter since it's used both in shared and non-shared modes
* [ ] Consider possible code re-use between subprocess code and single process test pool
* [ ] Review tests

SneakyMax commented:
I wanted to try out this branch, so If you'd like some more anecdotal numbers (Node 11.9.0):

### Vue Project

We have a Vue project that has 180 tests over 101 files. We use the `t.snapshot` feature a lot:

- `ava` - 1:36, all tests pass
- `ava --worker-threads` - Cannot find any tests in any of the files.
- `ava --share-forks` - Hangs after 0:25 with 166 tests passed, 2 failed. Error in the Vue runtime file.
- `ava --single-process` - Exits after 0:28 with 68 tests passed, 0 failed. Same error as before, except this time it quits the whole process.

### Node Project

We have a GraphQL API that extensively clears the require cache during its tests, and all tests run serially in each file because they use global state. It has 1447 tests over 96 files:

- `ava` - 2:52, all tests pass
- `ava --worker-threads` - Cannot find any tests in any of the files
- `ava --share-forks` - Froze my VM
- `ava --single-process` - JavaScript heap out of memory after 333 tests

ijjk commented:
@SneakyMax thanks for trying out the branch! If you have a chance could you try running them again with the latest changes? 

I'm not able to reproduce the `--worker-threads` not being able to find tests error so any more info would be appreciated. I was able to reproduce the hanging and running out of memory errors and it appears there was a memory leak in the way I was clearing `require.cache` between test files which should be fixed now. 

You might need to clear the cache under `node_modules/.cache/ava` before trying again.

ijjk commented:
@novemberborn I understand it is a lot. I started looking over the checklist and tried addressing some of the points below.

> Review documentation (probably needs a dedicated section / recipe)

I can add more verbose info to the documentation just have been waiting until everythings been agreed on. 

> Bikeshed CLI flags

I was thinking it might be good to instead of having cli flags for each test pool to just have a test pool flag e.g. `--test-pool singleProcess` or `--test-pool sharedForks`. With the test pool flag we could also allow passing in a path to a custom test pool to achieve something like #1451 I'm open to thoughts on this though.

> "Shared fork" reusing fork leads to complicated code paths, both in the pool code and in the `Fork` class, the latter since it's used both in shared and non-shared modes

Are you thinking it would be better to have a seperate fork module for shared forks? I was mainly trying to avoid duplicating code. 

> Consider possible code re-use between subprocess code and single process test pool

I just pushed up some updates and was able to re-use the code in subprocess with single process test pool.

SneakyMax commented:
Sure thing @ijjk!

### Vue Project

- `ava` - 1:37, all tests pass
- `ava --worker-threads` - 1:49, all tests pass. Slower :(
- `ava --share-forks` - Hangs after 169/180 tests pass. A couple tests errored (`Error in nextTick: "TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'."`) (jsdom?)
- `ava --single-process` - Error after 68 tests pass: `node_modules/vue/dist/vue.runtime.common.js:7005
  var transitionDelays = styles[transitionProp + 'Delay'].split(', ');`

`--share-forks` is significantly slower than the previous time I tested, unfortunately. Whatever cache changes you made seem to have undone any performance benefit.

Looks like you fixed `--worker-threads` somehow though.

### Node Project

- `ava` - 2:43, all tests pass.
- `ava --worker-threads` - 3:14, all tests pass. Slower.
- `ava --share-forks` - VM froze after 196 tests passed.
- `ava --single-process` - Heap out of memory after 333 tests.

The out of memory problems might actually be that whatever require cache busting I'm using has a memory leak, and it never happens during normal ava because the worker processes shut down, taking the memory leaks with them.

ijjk commented:
@SneakyMax Interesting 🤔on the node project you might be right about having a memory leak in your require cache clearing also. I had to make sure to remove the module from the parent module's children to get it to free up. It's weird that the shared forks froze up faster than the single process though since the shared forks should have more memory available unless your environment has restricted memory so the shared forks hit it faster maybe.  

The Vue errors have me curious, I wonder if something is being modified globally that I'm not catching somehow. Is either of these able to be shared by any chance or possibly a minimum example of the set up so I can try investigating a little more?

ijjk commented on api.js:
```diff
@@ -204,43 +206,50 @@ class Api extends Emittery {
+							workerOptions.updateSnapshots = true;
+						}
+
+						this.Fork = Fork;
```
This is mostly for mocking during testing. I thought it would be easier only mock here instead of each `test-pool` that needs `Fork`

ijjk commented on api.js:
```diff
@@ -204,43 +206,50 @@ class Api extends Emittery {
-							});
-						}, {concurrency});
+						// Default to using a new fork for each test (provides most isolation)
+						let ProcessPool = ForkTestPool;
```
Yeah that's better 👍

ijjk commented on lib/cli.js:
```diff
@@ -38,6 +38,9 @@ exports.run = () => { // eslint-disable-line complexity
 		  --serial, -s            Run tests serially
 		  --concurrency, -c       Max number of test files running at the same time (Default: CPU cores)
+		  --single-process        Run tests in single process (not isolated)
+		  --share-forks           Run tests in shared forks
```
Yup, updated

ijjk commented on lib/cli.js:
```diff
@@ -38,6 +38,9 @@ exports.run = () => { // eslint-disable-line complexity
 		  --concurrency, -c       Max number of test files running at the same time (Default: CPU cores)
+		  --single-process        Run tests in single process (not isolated)
+		  --share-forks           Run tests in shared forks
+		  --worker-threads        Run tests in worker threads
```
Updated

ijjk commented on lib/cli.js:
```diff
@@ -208,7 +223,10 @@ exports.run = () => { // eslint-disable-line complexity
+		ranFromCli: true,
+		shareForks: conf.shareForks,
+		singleProcess: conf.singleProcess,
+		workerThreads: conf.workerThreads
```
Yeah that makes sense, combining them can add a lot of complexity

ijjk commented on lib/test-pools/single-process-test-pool.js:
```diff
@@ -0,0 +1,184 @@
+
+		const runner = new Runner({
+			file,
+			scheduledStart: true,
```
Updated

ijjk commented on lib/test-pools/single-process-test-pool.js:
```diff
@@ -0,0 +1,184 @@
+
+	_worker(file, options) {
+		// Clear any global variables from previous test
+		revertGlobal();
```
Updated

ijjk commented on lib/fork.js:
```diff
@@ -31,107 +31,157 @@ const describeTTY = tty => ({
+		this.finished = false;
+		this.forcedExit = false;
+		this.emitter = new Emittery();
+		this.promise = new Promise((resolve, reject) => {
```
This was named `promise` to match the current `fork.promise`. We can rename it just we would also have to update it in `./api.js` and everywhere else.

ijjk commented on lib/test-pools/shared-fork-test-pool.js:
```diff
@@ -0,0 +1,63 @@
+
+		// Wait for forks to be initialized
+		return Bluebird.all(this.forkPromises).then(() => {
+			return Bluebird.map(files, file => {
```
Yeah that makes sense, updated

ijjk commented on lib/test-pools/single-process-test-pool.js:
```diff
@@ -0,0 +1,184 @@
+			exit(1);
+		});
+
+		/* istanbul ignore next */
```
Removed all except those related to `worker_threads`, since I can't test them on unsupported node versions it brings down overall code coverage

novemberborn replied:
I'd have expected `codecov` to combine coverage reports across all the Node.js versions we test on. This isn't happening?

ijjk commented on lib/test-pools/single-process-test-pool.js:
```diff
@@ -0,0 +1,184 @@
+				const {children} = require.cache[file] || {};
+				if (children) {
+					const dependencies = children.map(mod => mod.id);
+					statusListener({type: 'dependencies', dependencies});
```
This has been updated now that I'm re-using subprocess for `SingleProcessTestPool`

ijjk commented on lib/test-pools/test-pool.js:
```diff
@@ -0,0 +1,99 @@
+   * @returns {Function} worker.exit - function to exit worker
+   * @returns {Function} worker.onStateChange - function called to add stateChange callback
+   * @returns {Function} worker.notifyOfPeerFailure - function to try to stop currently scheduled tests.
+   */
```
Not really, mainly there to be verbose. Removed.

ijjk commented on lib/worker/main.js:
```diff
@@ -1,5 +1,12 @@
-const runner = require('./subprocess').getRunner();
+// If getRunner is missing AVA was probably required
+// directly instead of in a test
+if (typeof global.getRunner === 'undefined') {
```
I updated this to only be used in `SingleProcess` mode to fix an issue with `esm` causing the `getRunner` export not to be set here.

SneakyMax commented:
@ijjk sorry about the late reply - unfortunately I can't share either of the projects :(, I'm happy to keep testing the branch though. I hope we can get the speedup I first experienced, without any errors.

ijjk commented:
@SneakyMax I understand. I added an option `cacheRequire` that defaults to `true` which allows you to either require `options.require` everytime before new tests or to only require them once before all tests in that process. 

Note: this is only used in `sharedForks` and `singleProcess` mode. This might help increase performance a bit if your tests are set up to be able to allow `cacheRequire` to be enabled.

Retsam commented on lib/revert-global.js:
```diff
@@ -0,0 +1,47 @@
+	const storedKeys = {};
+	const storedCache = Object.assign({}, require.cache);
+	const storedExtensions = Object.assign({}, module.constructor._extensions);
+	const preventCache = ['bluebird', '../index.js', './worker/main.js'];
```
Why is Bluebird being prevented from caching? 

I've tried running my tests against this branch, and all of our tests that involve bluebird cancellation are failing: we currently have a 'testBootstrap.js" file the we import before all of our tests which enables bluebird cancellation, but it's not being respected, likely due to this logic here.

Retsam replied:
Perhaps `.getNewLibraryCopy()` would be useful so that the test framework can have its own copy of Bluebird without stepping on the application code?

ijjk replied:
I was clearing bluebird there to try to not interfere with the test's usage of bluebird but you might be right that clearing it there ended up interfering anyways. I moved it to only clear initially so hopefully it doesn't conflict now.

Retsam replied:
Yup, it looks like that fixed it.  All of my cancellation tests are running correctly now.

novemberborn commented:
Hi @ijjk, sorry I haven't gotten round to picking this up again.

ijjk commented:
@novemberborn I just updated the branch to resolve conflicts, it should be good to review when you have a chance.

thasmo commented:
Awesome work @ijjk! Hope this gets reviewed soon. 👍

novemberborn commented:
> Hope this gets reviewed soon. 👍

Yes me too 😛 

This is the next big PR I'm meaning to look at once we land https://github.com/avajs/ava/pull/1947. Apologies for taking so long. It's really hard to find the time (several uninterrupted hours) to go through PRs with this kind of impact.

novemberborn commented:
Once more, apologies for taking so long in looking at this.

The bad news is, this is still a really large PR! But, the good news is that we're now supporting experimental features. See this commit for an example: https://github.com/avajs/ava/pull/1947/commits/2b8ba3ac99f7ed9ea6ea5297c915a16bbcf28a17

@ijjk if you're still keen, perhaps you could make a PR for just one of the options we discussed here. Preferably not anything that needs sandboxes and cleanups. If we put that behind a feature flag we can ship it as a new, opt-in feature without being a breaking change. We can then build on that feature, one PR at a time. It'll be a lot easier to review multiple smaller changes than one big one. We'll also be free to change the implementation as we learn more.

If you're burned out by this experience I totally understand, and again I'm sorry it got to this point. Let us know and perhaps somebody else is interested in picking this up.

This PR is still too daunting for me so I'm going to go ahead and close it. But I'd really like to see these new worker modes 😍
